### DISCUSSION TITLE
batch update in jotai?

### DISCUSSION THREAD

**Discussion by yassuge** [pmndrs/jotai#2416] [Q&A] [answered]
Hi everyone,

I've been searching through the discussions but haven't found a specific answer to my scenario, so I thought I'd ask here.

I'm working with a derived atom (y) that depends on multiple atoms from an atomFamily (let's say x1 to xn). In the getter function of atom y, I use get to retrieve values from this atomFamily with various parameters (param_1,....param_n).

I also have a worker that updates each of these atomFamily atoms individually. However, every single update triggers a re-computation of the derived atom y, which involves some heavy calculations. This is leading to performance issues.

Is there a way to batch these updates so that the derived atom y only re-computes once after all the relevant atomFamily atoms have been updated? Ideally, I'd like y to update only once per batch rather than on each individual atom update.

All of my atoms, including the derived one, are defined outside of any React components. In my React component, I'm using useAtomValue(y) to access the derived atom's value.

Thanks in advance for any guidance or solutions you can share!

**Comment by dai-shi** [comment]
Jotai works everything synchronously in the core.
So, in short, there's no direct solution to your scenario.

There would be some workarounds, but if you are using atomFamily, the write will be completely separated, so the easiest solution would be to make derived atom "async" and control the pending state on your own. Let me know if it sounds a viable approach.

**Comment by yassuge** [comment]
HI @dai-shi ,

Thank you for the prompt response.
For now I m using an update lock, basically, before starting to call set on each atomFamily atom (for loop), I set a refreshAtom to false, which is used in the derived atom to choose whether to perform calc or render the existing. After the for loop, I set the refreshAtom to true, which then triggers the derived atom to calc, ie once all the atomFamily atoms are updated. Is this a pattern you recommend, or you think of a better one?

Cheers!

**Reply by dai-shi** [reply, marked as answer]
Yes, that's exactly one way of what I mean by your own pending state.

**Reply by yassuge** [reply]
Great, thanks a lot @dai-shi
